crates/iceberg-rest-server/src/api/v1/warehouse.rs

```diff
@@ -75,6 +105,113 @@ pub trait WarehouseService<C: Catalog, A: AuthZHandler, S: SecretStore> {
+        let warehouse_id = &WarehouseIdent::from(request.warehouse_id);
+        let mut transaction = C::Transaction::begin_write(context.v1_state.catalog.clone()).await?;
+        let old_warehouse_config =
+            C::get_warehouse(warehouse_id, context.v1_state.catalog.clone()).await?;
+        let (old_name, old_storage_profile, old_storage_credential_id) = {
+            (
+                old_warehouse_config.warehouse_name,
+                old_warehouse_config.storage_profile,
+                old_warehouse_config.storage_credential_id,
+            )
+        };
+
+        if let Some(new_name) = request.new_name {
+            if new_name != old_name {
+                C::update_warehouse_name(warehouse_id, &new_name, transaction.transaction())
+                    .await?;
+            }
+        }
+
+        match (request.new_storage_credential, request.new_storage_profile) {
+            (None, Some(_)) => {
+                let err: IcebergErrorResponse = ErrorModel::builder()
+                    .message("Cannot update 'storage_profile' without new creds.")
+                    .code(StatusCode::CONFLICT.into())
+                    .r#type("CannotUpdateWarehouse")
+                    .build()
+                    .into();
+
+                Err(err)
+            }
+            (None, None) => Ok(()),
+            (Some(new_storage_credential), Some(new_storage_profile)) => {
+                let mut new_storage_profile = new_storage_profile;
+                new_storage_profile
+                    .validate(Some(&new_storage_credential))
+                    .await?;
+
+                let can_be_updated = match (&new_storage_profile, old_storage_profile) {
+                    (StorageProfile::S3(new_profile), StorageProfile::S3(old_profile)) => {
+                        old_profile.can_be_updated_with(new_profile)
+                    }
+                };
+
+                if can_be_updated {
+                    C::update_warehouse_storage_profile(
+                        warehouse_id,
+                        new_storage_profile,
+                        transaction.transaction(),
+                    )
+                    .await?;
+                }
+
+                if let Some(old_cred) = old_storage_credential_id {
+                    S::delete_secret(&old_cred, context.v1_state.secrets.clone()).await?;
+                }
+
+                let new_secret_id =
+                    Some(S::create_secret(new_storage_credential, context.v1_state.secrets).await?);
+
+                Ok(C::update_warehouse_storage_secret_id(
+                    warehouse_id,
+                    new_secret_id,
+                    transaction.transaction(),
+                )
+                .await?)
+            }
+            (Some(new_storage_credential), None) => {
+                if let Some(old_cred) = old_storage_credential_id {
+                    S::delete_secret(&old_cred, context.v1_state.secrets.clone()).await?;
```

Comment:
This also needs to be at the very end after the transaction succeeded.
Also use .ok() here and just log an error. If this fails, the update still works. We need an external reconciliation process later.